pkg/gui/controllers/local_commits_controller.go

```diff
@@ -839,37 +843,88 @@ func (self *LocalCommitsController) squashFixupCommits() error {
+	return self.squashFixupsImpl(commit, rebaseStartIdx)
+}
+
+func (self *LocalCommitsController) squashFixupsImpl(commit *models.Commit, rebaseStartIdx int) error {
+	selectionOffset := countSquashableCommitsAbove(self.c.Model().Commits, self.context().GetSelectedLineIdx(), rebaseStartIdx)
+	return self.c.WithWaitingStatusSync(self.c.Tr.SquashingStatus, func() error {
 		self.c.LogAction(self.c.Tr.Actions.SquashAllAboveFixupCommits)
 		err := self.c.Git().Rebase.SquashAllAboveFixupCommits(commit)
-		return self.c.Helpers().MergeAndRebase.CheckMergeOrRebase(err)
+		self.context().MoveSelectedLine(-selectionOffset)
+		return self.c.Helpers().MergeAndRebase.CheckMergeOrRebaseWithRefreshOptions(
+			err, types.RefreshOptions{Mode: types.SYNC})
 	})
 }
 
-func (self *LocalCommitsController) findCommitForSquashFixupsInCurrentBranch() (*models.Commit, error) {
+func (self *LocalCommitsController) findCommitForSquashFixupsInCurrentBranch() (*models.Commit, int, error) {
 	commits := self.c.Model().Commits
 	_, index, ok := lo.FindIndexOf(commits, func(c *models.Commit) bool {
 		return c.IsMerge() || c.Status == models.StatusMerged
 	})
 
 	if !ok || index == 0 {
-		return nil, errors.New(self.c.Tr.CannotSquashCommitsInCurrentBranch)
+		return nil, -1, errors.New(self.c.Tr.CannotSquashCommitsInCurrentBranch)
+	}
+
+	return commits[index-1], index - 1, nil
+}
+
+// Anticipate how many commits above the selectedIdx are going to get squashed
+// by the SquashAllAboveFixupCommits call, so that we can adjust the selection
+// afterwards. Let's hope we're matching git's behavior correctly here.
+func countSquashableCommitsAbove(commits []*models.Commit, selectedIdx int, rebaseStartIdx int) int {
+	result := 0
+
+	// For each commit _above_ the selection, ...
+	for i, commit := range commits[0:selectedIdx] {
+		// ... see if it is a fixup commit, and get the base subject it applies to
+		if baseSubject, isFixup := isFixupCommit(commit.Name); isFixup {
+			// Then, for each commit after the fixup, up to and including the
+			// rebase start commit, see if we find the base commit
+			for _, baseCommit := range commits[i+1 : rebaseStartIdx+1] {
+				if strings.HasPrefix(baseCommit.Name, baseSubject) {
+					result++
+				}
+			}
+		}
+	}
+	return result
+}
+
+// Check whether the given subject line is the subject of a fixup commit, and
+// returns (trimmedSubject, true) if so (where trimmedSubject is the subject
+// with all fixup prefixes removed), or (subject, false) if not.
+func isFixupCommit(subject string) (string, bool) {
+	prefixes := []string{"fixup! ", "squash! ", "amend! "}
+	trimPrefix := func(s string) (string, bool) {
+		for _, prefix := range prefixes {
+			if strings.HasPrefix(s, prefix) {
+				return strings.TrimPrefix(s, prefix), true
+			}
+		}
+		return s, false
+	}
+
+	if subject, wasTrimmed := trimPrefix(subject); wasTrimmed {
+		for {
+			if subject, wasTrimmed = trimPrefix(subject); !wasTrimmed {
```

Comment:
I assume this for loop is here because of recursive fixups e.g. 'fixup! fixup! foo'. If so it would be good to leave a comment to that effect.

Out of curiosity, how granular are we with this logic i.e. if I have:
* fixup! fixup! foo
* fixup! foo
* bar
* foo

And I want to squash all fixups above bar, I'm pretty sure we'll end up with:
* fixup! foo
* bar
* foo

Does this code handle that? For the record I don't actually care because it's such an absurd edge case and handling it would likely require more complex code but I can't quite tell from just looking at the code.

Reply:
> I assume this for loop is here because of recursive fixups e.g. 'fixup! fixup! foo'. If so it would be good to leave a comment to that effect.

Added a comment in cd75fbce0404.

> Out of curiosity, how granular are we with this logic i.e. if I have:
> 
> * fixup! fixup! foo
> * fixup! foo
> * bar
> * foo
> 
> And I want to squash all fixups above bar, I'm pretty sure we'll end up with:
> 
> * fixup! foo
> * bar
> * foo

No, you'll end up with
* bar
* foo

Git takes care of folding recursive prefixes, and I replicated this logic here.

> Does this code handle that? For the record I don't actually care because it's such an absurd edge case

Not absurd at all, in fixup-heavy workflows it's a very common thing in practice. We have it at work all the time.

Reply:
> Not absurd at all, in fixup-heavy workflows it's a very common thing in practice. We have it at work all the time.

Really? Where you only want to squash some fixup commits but not others, and some of those are recursive? Interesting.

Reply:
I'm not sure what you mean by "only want to squash some fixup commits but not others"? We usually want to squash them all.

Reply:
Oh wait, I misread your scenario. I overlooked the "above bar". (It's early in the morning, I didn't have coffee yet.)

In that case you'll end up with 
```
* fixup! fixup! foo
* fixup! foo
* bar
```
which is the expected outcome, and my code handles this correctly.